Tests/Algorithm/AlgorithmSetHoldingsTests.cs

```diff
@@ -0,0 +1,220 @@
+                    finalPositions.Cast<object>().ToList(),
+                    feeTypes.Cast<object>().ToList(),
+                    priceMovements.Cast<object>().ToList(),
+                    leverages.Cast<object>().ToList()
+                };
+                var permutations = new List<object[]>();
+                Permuter<object>.Permute(data, permutations);
+
+                var ret = permutations
+                    .Where(row => (Position)row[0] != (Position)row[1])     // initialPosition != finalPosition
+                    .Select(row => new TestCaseData(row).SetName(string.Join("_", row)))
+                    .ToArray();
+
+                return ret;
+            }
+        }
+
+        [Test, TestCaseSource("TestParameters")]
+        public void Run(Position initialPosition, Position finalPosition, FeeType feeType, PriceMovement priceMovement, int leverage)
+        {
+            Console.WriteLine("----------");
+            Console.WriteLine("PARAMETERS");
+            Console.WriteLine("Initial position: " + initialPosition);
+            Console.WriteLine("Final position: " + finalPosition);
+            Console.WriteLine("Fee type: " + feeType);
+            Console.WriteLine("Price movement: " + priceMovement);
+            Console.WriteLine("Leverage: " + leverage);
+            Console.WriteLine("----------");
+            Console.WriteLine();
+
+            var algorithm = new QCAlgorithm();
+
+            var security = algorithm.AddSecurity(_symbol.ID.SecurityType, _symbol.ID.Symbol);
+            security.FeeModel = _feeModels[feeType];
+            security.SetLeverage(leverage);
+
+            algorithm.SetCash(Cash);
+
+            Update(security, BasePrice);
+
+            decimal targetPercentage;
+            OrderDirection orderDirection;
+            MarketOrder order;
+            decimal orderFee;
+            OrderEvent fill;
+            int orderQuantity;
+            decimal freeMargin;
+            decimal requiredMargin;
+            if (initialPosition != Position.Zero)
+            {
+                targetPercentage = (decimal)initialPosition;
+                orderDirection = initialPosition == Position.Long ? OrderDirection.Buy : OrderDirection.Sell;
+                orderQuantity = algorithm.CalculateOrderQuantity(_symbol, targetPercentage);
+                order = new MarketOrder(_symbol, orderQuantity, DateTime.UtcNow);
+                freeMargin = algorithm.Portfolio.GetMarginRemaining(_symbol, orderDirection);
+                requiredMargin = security.MarginModel.GetInitialMarginRequiredForOrder(security, order);
+
+                Console.WriteLine("Current price: " + security.Price);
+                Console.WriteLine("Target percentage: " + targetPercentage);
+                Console.WriteLine("Order direction: " + orderDirection);
+                Console.WriteLine("Order quantity: " + orderQuantity);
+                Console.WriteLine("Free margin: " + freeMargin);
```

Comment:
we may want to comment out this output, when running the automated tests it will spam the logs on travis